### ISSUE TITLE
MavenBackingRepository.getMetadata doesn't do housekeeping

### ISSUE THREAD

**Issue by mnlipp** [bndtools/bnd#2263]
This is against 3.5.0.REL, but from the sources, the problem should still exist in master.

MavenBackingRepository maintains its own `maven-metadata-<id>.xml` file in the local repository. When `MavenBackingRepository.getMetadata` is called, it downloads the current version of `maven-metadata.xml` from the server into this file. It does *not*, however, check if the metadata indicates an update since the artifact has been downloaded (e.g. by comparing `RevisionMetadata.lastUpdated` from the `maven-metadata-<id>.xml` before download and after download or comparing with file dates, see below).

This causes existing `*-SNAPSHOT.{jar,pom}` files to be outdated without notice. The metadata is updated and the program works with up-to-date metadata, but when accessing the archive or pom files, you get the versions that have previously been downloaded, i.e. versions that don't match the metadata (unless they are older than one day, which forces a download -- but this certainly does not match the nature of snapshots).

IMHO, the `*-SNAPSHOT.{jar,pom}` should be deleted, if an update of the metadata is detected. I currently work around the problem like this:

```java
if (archive.isSnapshot()) {
	for (MavenBackingRepository mbr : mavenRepository.getSnapshotRepositories()) {
		if (mbr.getVersion(archive.getRevision()) != null) {
			// Found backing repository
			File metaFile = mavenRepository.toLocalFile(revision.metadata(mbr.getId()));
			RevisionMetadata metaData = MetadataParser.parseRevisionMetadata(
					metaFile);
			File archiveFile = mavenRepository.toLocalFile(archive);
			if (archiveFile.lastModified() < metaData.lastUpdated) {
				archiveFile.delete();
			}
			File pomFile = mavenRepository.toLocalFile(archive.getPomArchive());
			if (pomFile.lastModified() < metaData.lastUpdated) {
				pomFile.delete();
			}
			break;
		}
	}
}
```
(Taken from [here](https://github.com/mnlipp/de.mnl.osgi/blob/e849d92bf3ba5a197b62bf39a10cefdbf6e2012b/de.mnl.osgi.bnd.repository/src/de/mnl/osgi/bnd/repository/maven/provider/NexusSearchOsgiRepository.java#L349), in case more context is needed.)

This, or something with a similar effect, should be done in `MavenBackingRepository.getMetadata` (assuming that this is the only place where metadata is updated).

**Comment by rotty3000** [comment]
Good catch! I've been meaning to fix this bug but haven't had time to debug it. We had this issue about it already #2135

**Comment by bjhargrave** [comment]
Duplicate of #2135.

**Comment by rotty3000** [comment]
I'll try to fix this for 4.0
